File: cli/toob/commands/chip.py
"""``toob chip`` sub-commands — add, spawn, remove, list, info."""

from __future__ import annotations

import click

from toob.core.installer import ChipInstaller
from toob.core.paths import find_project_root
from toob.core.registry_cache import RegistryCache


@click.group()
def chip() -> None:
    """Manage chip HAL packages."""


@chip.command(name="list")
def list_chips() -> None:
    """List all chips available in the registry."""
    cache = RegistryCache()
    try:
        chips = cache.list_chips()
    except FileNotFoundError as exc:
        raise click.ClickException(str(exc)) from exc

    if not chips:
        click.echo("Registry is empty.")
        return

    click.echo(f"{'Chip':<20} {'Vendor':<12} {'Arch':<16} {'Version':<10}")
    click.echo("-" * 58)
    for name in sorted(chips):
        ci = chips[name]
        click.echo(f"{ci.name:<20} {ci.vendor:<12} {ci.arch:<16} {ci.version:<10}")


@chip.command()
@click.argument("name")
def info(name: str) -> None:
    """Show detailed information about a chip."""
    cache = RegistryCache()
    try:
        ci = cache.get_chip(name)
    except (FileNotFoundError, KeyError) as exc:
        raise click.ClickException(str(exc)) from exc

    click.echo(f"  Name:          {ci.name}")
    click.echo(f"  Version:       {ci.version}")
    click.echo(f"  Vendor:        {ci.vendor}")
    click.echo(f"  Architecture:  {ci.arch}")
    click.echo(f"  Toolchain:     {ci.toolchain}")
    click.echo(f"  Description:   {ci.description or '(none)'}")
    click.echo(f"  Min Toob:      {ci.min_toob_version}")


@chip.command()
@click.argument("name")
def add(name: str) -> None:
    """Install a chip HAL from the registry."""
    try:
        root = find_project_root()
        cache = RegistryCache()
        installer = ChipInstaller(root, cache)
        installer.add(name)
    except (FileNotFoundError, KeyError, RuntimeError) as exc:
        raise click.ClickException(str(exc)) from exc


@chip.command()
@click.argument("name")
def spawn(name: str) -> None:
    """Install a chip HAL as locally editable (tracked by git)."""
    try:
        root = find_project_root()
        cache = RegistryCache()
        installer = ChipInstaller(root, cache)
        installer.spawn(name)
    except (FileNotFoundError, KeyError, RuntimeError) as exc:
        raise click.ClickException(str(exc)) from exc


@chip.command()
@click.argument("name")
def remove(name: str) -> None:
    """Uninstall a chip HAL and its unshared dependencies."""
    try:
        root = find_project_root()
        cache = RegistryCache()
        installer = ChipInstaller(root, cache)
        installer.remove(name)
    except (FileNotFoundError, KeyError, RuntimeError) as exc:
        raise click.ClickException(str(exc)) from exc

File: cli/toob-cli/cmd/chip.go
package cmd

import (
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"github.com/toob-boot/toob/internal/installer"
	"github.com/toob-boot/toob/internal/paths"
	"github.com/toob-boot/toob/internal/registry"
	"github.com/toob-boot/toob/internal/ui"
)

var chipCmd = &cobra.Command{
	Use:   "chip",
	Short: "Manage chip HAL packages",
}

var chipListCmd = &cobra.Command{
	Use:   "list [query]",
	Short: "List all chips available in the registry",
	Args:  cobra.MaximumNArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		query := ""
		if len(args) > 0 {
			query = strings.ToLower(args[0])
		}

		cache := registry.NewCache("")

		// Async Matrix Fetch
		matrixChan := make(chan *registry.Matrix, 1)
		go func() {
			m, _ := cache.FetchLiveMatrix()
			matrixChan <- m
		}()

		idx, err := cache.LoadIndex()
		if err != nil {
			return err
		}
		if len(idx.Chips) == 0 {
			ui.Muted("Registry is empty.")
			return nil
		}

		// Wait up to 2 seconds for matrix
		var matrix *registry.Matrix
		select {
		case matrix = <-matrixChan:
		case <-time.After(2 * time.Second):
		}

		ui.Header("Chip Registry")

		headers := []string{"Chip", "Version", "Vendor", "Arch", "Verified CLI"}
		var rows [][]string
		for _, ci := range idx.Chips {
			if query != "" && !strings.Contains(strings.ToLower(ci.Name), query) {
				continue
			}

			verifiedClisStr := ui.Yellow("unverified")
			if matrix != nil {
				if chipEntry, has := (*matrix)[ci.Name]; has {
					searchVer := strings.TrimPrefix(ci.Version, "v")
					for vKey, verEntry := range chipEntry.Versions {
						if strings.TrimPrefix(vKey, "v") == searchVer {
							var clis []string
							for cliVer, info := range verEntry.VerifiedCliVersions {
								if info.Status == "VERIFIED" {
									clis = append(clis, cliVer)
								}
							}
							if len(clis) > 0 {
								verifiedClisStr = ui.Green(strings.Join(clis, ", "))
							}
							break
						}
					}
				}
			}

			rows = append(rows, []string{ui.Bold(ci.Name), ui.Brand(ci.Version), ci.Vendor, ci.Arch, verifiedClisStr})
		}

		if len(rows) == 0 {
			ui.Warn("No chips found matching '%s'", query)
			return nil
		}

		ui.Table(headers, rows)
		ui.Muted("%d chip(s) available.", len(rows))
		return nil
	},
}

var chipInfoCmd = &cobra.Command{
	Use:   "info [name]",
	Short: "Show detailed information about a chip",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		cache := registry.NewCache("")
		ci, err := cache.GetChip(args[0])
		if err != nil {
			return err
		}

		ui.Header(fmt.Sprintf("Chip: %s", ci.Name))

		if ci.Verified {
			ui.Success("CI Matrix: Verified")
		} else {
			ui.Warn("CI Matrix: Unverified")
		}

		idx, _ := cache.LoadIndex()
		vVer := "unknown"
		aVer := "unknown"
		if idx != nil {
			if vInfo, ok := idx.Vendors[ci.Vendor]; ok {
				vVer = vInfo.Version
			}
			if aInfo, ok := idx.Archs[ci.Arch]; ok {
				aVer = aInfo.Version
			}
		}

		fmt.Fprintln(os.Stderr)
		ui.KeyValue("Version", ui.BoldBrand(ci.Version))
		ui.KeyValue("Vendor", fmt.Sprintf("%s %s", ci.Vendor, ui.Gray("(v"+vVer+")")))
		ui.KeyValue("Architecture", fmt.Sprintf("%s %s", ci.Arch, ui.Gray("(v"+aVer+")")))
		ui.KeyValue("Compiler", ui.Cyan(ci.CompilerPrefix))
		ui.KeyValue("Registry Path", ui.Gray(ci.Path))
		if ci.Description != "" {
			ui.KeyValue("Description", ci.Description)
		}
		ui.Divider()
		return nil
	},
}

var chipAddCmd = &cobra.Command{
	Use:   "add [name]",
	Short: "Install a chip HAL from the registry",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		root, err := paths.FindProjectRoot("")
		if err != nil {
			return err
		}
		cache := registry.NewCache("")
		inst := installer.New(root, cache)
		return inst.Add(args[0])
	},
}

var chipSpawnCmd = &cobra.Command{
	Use:   "spawn [name]",
	Short: "Install a chip HAL as locally editable (tracked by git)",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		root, err := paths.FindProjectRoot("")
		if err != nil {
			return err
		}
		cache := registry.NewCache("")
		inst := installer.New(root, cache)
		return inst.Spawn(args[0])
	},
}

var chipRemoveCmd = &cobra.Command{
	Use:   "remove [name]",
	Short: "Uninstall a chip HAL and its unshared dependencies",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		root, err := paths.FindProjectRoot("")
		if err != nil {
			return err
		}
		cache := registry.NewCache("")
		inst := installer.New(root, cache)
		return inst.Remove(args[0])
	},
}

func init() {
	chipCmd.AddCommand(chipListCmd)
	chipCmd.AddCommand(chipInfoCmd)
	chipCmd.AddCommand(chipAddCmd)
	chipCmd.AddCommand(chipSpawnCmd)
	chipCmd.AddCommand(chipRemoveCmd)
}
